Implement the ability to send metrics via UDS

s-vitaliy opened the pull request:
Resolves #67


This pull request implements the ability to send ZIO metrics using the unix domain socket.

## Implemented:

**NOTE** The original implementation was changed based on the reviewer comments, see notes in the [comment](https://github.com/zio/zio-metrics-connectors/pull/70#issuecomment-2916672917) below.

- ~The `StatsdClient` class was replaced with a new class, `DogStatsdClient`, which was generalized to enable it to send data using UDS.~
- ~A new configuration, `DatadogUdsConfig`, was added, which can be used to set up the DataDog client to use UDS.~
- ~Unit tests were added.~

**NOTE 1:** ~The name `DogStatsdClient` was chosen because DogStatsd is the name of [the StatsD protocol extension](https://docs.datadoghq.com/developers/dogstatsd/) used by DataDog.~

**NOTE 2:** ~The DogStatsdClient, similar to the [previously used StatsdClient class](https://github.com/zio/zio-metrics-connectors/blob/09314e4870ff0684ef8ddd162a624edb9a1c7467/statsd/src/main/scala/zio/metrics/connectors/statsd/StatsdClient.scala#L24) in the DataDog connector implementation, does not include proper error handling. However, this does not impact the current functionality and the need for error handling is beyond the scope of this Pull Request.~

s-vitaliy committed:
Implement the ability to send metrics via UDS

s-vitaliy committed:
Self-review fixes

s-vitaliy committed:
Fix formatting issues

s-vitaliy committed:
Add motivation to docstrings

s-vitaliy committed:
Also add some comments

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DogStatsdClient.scala:
```diff
@@ -0,0 +1,73 @@
+
+trait DogStatsdClient {
+  private[connectors] def send(chunk: Chunk[Byte]): Long
+}
```
I don't think we need to introduce a new trait for this. It can be a final class that extends `StatsdClient`

s-vitaliy replied:
Removed this interface.

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DataDogEventProcessor.scala:
```diff
@@ -11,7 +11,7 @@ import zio.metrics.connectors.statsd.StatsdClient
 
   def make(
-    client: StatsdClient,
+    client: DogStatsdClient,
```
We should keep this as `StatsdClient` and have `DogStatsdClient` extend `StatsdClient` (as per my comment above)

s-vitaliy replied:
Fixed.

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DatadogConfig.scala:
```diff
@@ -5,7 +5,19 @@ import java.time.Duration
+  val entityId: Option[String]
+  val maxQueueSize: Int
+  val sendUnchanged: Boolean
+}
```
Unfortunately this is not a binary nor source compatible change. We could bump the major version of the library but I'm not sure if this change is enough to warrant it.

In either way I'm not too happy with the current design of the `datadogLayer` and I think we should do the following:
1. Deprecate `DatadogConfig`
2. Add `DatadogPublisherConfig` that has all the fields that are in this trait minus network-specific ones
3. Deprecate `datadogLayer`
4. Add a new layer that has this signature: `ZLayer[DatadogPublisherConfig && StatsdClient & MetricsConfig, Nothing, Unit]`
5. Users can then select their preferred `StatsdClient` from the `statsd` package

s-vitaliy replied:
Applied this suggestion in the new commit.

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DogStatsdClient.scala:
```diff
@@ -0,0 +1,73 @@
+    private[connectors] def write(byteBuffer: ByteBuffer): Try[Long] = Try(channel.write(byteBuffer).toLong)
+  }
+
+  private class UdsWriter(channel: DatagramChannel, address: UnixSocketAddress) extends DogStatsdSocketWriter {
```
You mentioned on Discord that this is a datadog-specific implementation, but AFAICT it should be compatible with any agent that supports UDS right? I think we might be better off moving this implementation to the statsd package?

s-vitaliy replied:
Not exactly, it should be compatible with any datagram UDS. For instance, the DataDog agent creates the datagram UDS for metrics, but it creates the stream socket for traces.

However, you are probably right. If any other agent creates the datagram UDS, it should also be compatible.

kyri-petrou reviewed (commented):
Thanks a lot for this! I think the implementation overall looks good to me, but I have some concerns around UX / API semantics (see my comments below). We can discuss this more on Discord if any of my comments are unclear.

Also, we need to make the corresponding changes to the documentation so that users are aware of this new functionality, along with pros/cons of each approach. From datadog's documentation it seems that support for UDS was added in agent version 6+ and that it's available only on Linux so maybe that's another thing we need to point out

s-vitaliy committed:
Address the comments from the code review

s-vitaliy committed:
Fix incorrect client

s-vitaliy committed:
Fix formatting issues

s-vitaliy committed:
Fix type errors

s-vitaliy committed:
Forgot to rename writer

s-vitaliy commented:
@kyri-petrou, I've updated this pull request accordingly to your comments. Here what's changed:
- `DogStatsdClient` was moved to `statsd` package and renamed to `DatagramSocketClient` since it's more generic and can be used with any agent that uses the datagram socket protocol.
- `DatadogConfig` was deprecated and replaced with `DatadogPublisherConfig` case class.
- Some **private** functions in the `datadog` object that were using `DatadogConfig` were updated to use `DatadogPublisherConfig` instead.
- I removed the ability to send events over network from the `DatagramSocketClient` since it can be used with any agent that uses the datagram socket protocol.
- I added the `DatagramSocketConfig` to the `statsd` package to store the configuration for the `DatagramSocketClient`.

I **did not** update the documentation yet, since I wanted to get your feedback first.
If you are ok with these changes, I will update the documentation accordingly (in this PR or in a separate one, depending on your preference).

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DatadogEncoder.scala:
```diff
@@ -9,13 +9,13 @@ case object DatadogEncoder {
   private val BUF_PER_METRIC = 128
 
-  def encoder(config: DatadogConfig): MetricEvent => Task[Chunk[Byte]] = {
+  def encoder(config: DatadogPublisherConfig): MetricEvent => Task[Chunk[Byte]] = {
```
Keep the original methods to keep the code binary compatible. You can overload it to accept `DatadogConfig`, deprecate it, and within it map it to `DatadogPublisherConfig` and pass it to the new method

s-vitaliy replied:
Done

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DatadogEncoder.scala:
```diff
@@ -9,13 +9,13 @@ case object DatadogEncoder {
 
   def histogramEncoder(
-    config: DatadogConfig,
+    config: DatadogPublisherConfig,
```
Same as above

kyri-petrou commented on statsd/src/main/scala/zio/metrics/connectors/statsd/DatagramSocketClient.scala:
```diff
@@ -0,0 +1,50 @@
+                 })
+    } yield new UdsWriter(channel, new UnixSocketAddress(path))
+
+  private[connectors] def make: ZIO[Scope & DatagramSocketConfig, Nothing, StatsdClient] =
```
We're going to need to create a layer that points to this method otherwise users won't be able to create the `StatsdClient` for UDS.

We'll need to place this layer in the statsd package object AFAICT

s-vitaliy replied:
Added the layer

kyri-petrou reviewed (commented):
LGTM, just some minor comments below 🙏

s-vitaliy committed:
Address the reviewer's comments

s-vitaliy committed:
Update documentation

s-vitaliy committed:
Fix formatting issues

s-vitaliy committed:
Self-review fixes

s-vitaliy committed:
Self-review fixes

s-vitaliy commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DataDogEventProcessor.scala:
```diff
@@ -13,15 +13,15 @@ object DataDogEventProcessor {
     queue: RingBuffer[(MetricKey[MetricKeyType.Histogram], Double)],
-  ): ZIO[DatadogConfig & MetricsConfig, Nothing, Unit] =
+    datadogPublisherConfig: DatadogPublisherConfig,
+  ): ZIO[MetricsConfig, Nothing, Unit] =
```
You didn't mention this method in the review, but I created an overload that accepts `DatadogPublisherConfig` and marked the method that accepts `DatadogConfig` as deprecated to be on the safe side 🙂  (see below).

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/DatadogEncoder.scala:
```diff
@@ -27,7 +37,11 @@ case object DatadogEncoder {
     (key, values) => Chunk.fromArray(encodeHistogramValues(key, values).toString().getBytes())
   }
 
+  @deprecated("Use the overload that accepts DatadogPublisherConfig instead", "2.4.0")
   private def makeStatsdEncoder(config: DatadogConfig): StatsdEncoder =
+    makeStatsdEncoder(DatadogConfig.toPublisherConfig(config))
```
This is a private method. We don't need to worry about binary compatibility in private methods so you can simply delete it

kyri-petrou commented on datadog/src/main/scala/zio/metrics/connectors/datadog/package.scala:
```diff
@@ -25,17 +26,44 @@ package object datadog {
+
+  @deprecated("Use the overload that accepts DatadogPublisherConfig instead", "2.4.0")
   private def eventFilter(config: DatadogConfig): MetricEvent => Boolean =
+    eventFilter(DatadogConfig.toPublisherConfig(config))
```
As above, you can safely delete this method

kyri-petrou commented on statsd/src/main/scala/zio/metrics/connectors/statsd/DatagramSocketClient.scala:
```diff
@@ -0,0 +1,50 @@
+    override def send(chunk: Chunk[Byte]): Long =
+      writer.write(ByteBuffer.wrap(chunk.toArray)) match {
+        case Success(value) =>
+          // println(s"Sent UDP data [$value]")
```
Remove commented code

kyri-petrou commented on statsd/src/main/scala/zio/metrics/connectors/statsd/DatagramSocketClient.scala:
```diff
@@ -0,0 +1,50 @@
+          // println(s"Sent UDP data [$value]")
+          value
+        case Failure(_)     =>
+          // t.printStackTrace()
```
As above

kyri-petrou commented on statsd/src/main/scala/zio/metrics/connectors/statsd/package.scala:
```diff
@@ -5,11 +5,28 @@ import zio.metrics.connectors.internal.MetricsClient
 
 package object statsd {
 
+  @deprecated("Use zio.metrics.connectors.datadog.live instead", "2.4.0")
```
I think the deprecation message is wrong, it should point to the two new layers

kyri-petrou commented on statsd/src/main/scala/zio/metrics/connectors/statsd/package.scala:
```diff
@@ -5,11 +5,28 @@ import zio.metrics.connectors.internal.MetricsClient
       StatsdClient.make.flatMap(clt => MetricsClient.make(statsdHandler(clt))).unit,
     )
 
+  lazy val live: ZLayer[StatsdConfig & MetricsConfig, Nothing, StatsdClient] =
```
I think we. can be a bit smarter with the names here. One can be `statsdUDP` and another `statsdUDS` or something alone these lines.

We should also add some short scaladoc explaining their difference

s-vitaliy committed:
Review fixes

s-vitaliy committed:
Fix formatting error

kyri-petrou reviewed (approved):
Thanks!

nafg commented on statsd/src/main/scala/zio/metrics/connectors/statsd/package.scala:
```diff
@@ -5,11 +5,34 @@ import zio.metrics.connectors.internal.MetricsClient
+  /**
+   * Creates a layer that provides a StatsdClient that sends metrics over UDP network protocol.
+   */
+  lazy val statsdUDP: ZLayer[StatsdConfig & MetricsConfig, Nothing, StatsdClient] =
+    ZLayer.scoped {
+      for {
+        config <- ZIO.service[StatsdConfig]
+        client <- StatsdClient.make.provideSome[Scope](ZLayer.succeed(config))
+      } yield client
+    }
```
How is this code different from

```scala
ZLayer.scoped {
  StatsdClient.make
}
```
